dogstatsd/aggregator: add more memory telemetry

iksaif opened the pull request:
### What does this PR do?

Adds the following telemetry:

Packets:
* packets_channel_queued_packets: number of packets in the channel
by listener_id, if enabled
* packets_channel_size_packets: number of packets per listener-id
* packets_channel_size_packets_bytes: bytes in packets per listener-id
* packets_buffer_size_bytes: bytes per listener-id
* dogstatsd_contexts_bytes_by_mtype: context bytes per aggregator shard and metric type

Checks:
* checks_contexts
if explicitly enabled, by check ID
* checks_contexts_by_mtype
* checks_contexts_by_mtype
* checks_contexts_bytes_by_mtype

Tagset:
* tagset_sum_tags_bytes: bytes of the structure
* tagset_sum_tags_data_bytes: bytes of referenced tags (assuming not deduplicated)

### Motivation

We need a better understanding of the memory usage to make improvements. We also need to track more and more things per listener_id to be able to apply targeted back-pressure.

### Additional Notes

The high cardinality metrics are disabled by default

### Possible Drawbacks / Trade-offs

We might degraded performance so we will have to be careful about the regression tests

### Describe how to test/QA your changes

Start the agent, send a few dogstatsd metrics and validate that the new metrics are populated in localhost:5000/telemetry

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

iksaif committed:
dogstatsd/aggregator: add more memeory telemetry

iksaif committed:
Fix tests

iksaif committed:
Make the linter happy

iksaif committed:
Make the linter happier

pgimalac commented on pkg/util/size.go:
```diff
@@ -0,0 +1,188 @@
+	IntSize = strconv.IntSize / 8
+	// StringSize is the size of a string structure in bytes.
+	StringSize = unsafe.Sizeof("")
+	// StringSliceSice is the size of the string slice in bytes (not counting the size of the strings themselves).
+	StringSliceSice = unsafe.Sizeof([]string{})
+)
+
+// SizeOfString returns the size of a string structure in bytes.
+func SizeOfString() int {
+	return int(StringSize)
+}
+
+// SizeOfStringSlice returns the size of the string slice in bytes (not counting the size of the strings themselves).
+func SizeOfStringSlice(s []string) int {
+	return int(StringSliceSice) + len(s)*SizeOfString()
+}
```
```suggestion
	// StringSliceSize is the size of the string slice in bytes (not counting the size of the strings themselves).
	StringSliceSize = unsafe.Sizeof([]string{})
)

// SizeOfString returns the size of a string structure in bytes.
func SizeOfString() int {
	return int(StringSize)
}

// SizeOfStringSlice returns the size of the string slice in bytes (not counting the size of the strings themselves).
func SizeOfStringSlice(s []string) int {
	return int(StringSliceSize) + len(s)*SizeOfString()
}
```

pgimalac commented on pkg/util/size.go:
Why copy paste https://github.com/DmitriyVTitov/size/blob/master/size.go instead of just importing the dependency ?
If you want to keep this I believe we'll need to explicitly keep the original copyright with the author and the original license.

iksaif replied:
Good point, I ended up not using this part of the code  since I did most size manually, I'll just remove the code until we need it

iksaif committed:
Update pkg/util/size.go

Co-authored-by: Pierre Gimalac <pierre.gimalac@datadoghq.com>

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif committed:
Remove unused size utilities

vickenty commented on comp/dogstatsd/packets/telemetry.go:
```diff
@@ -15,7 +15,11 @@ var (
 	// tagging it with listener_id in case this changes later.
 	tlmChannelSize = telemetry.NewGauge("dogstatsd", "packets_channel_size",
+		[]string{}, "Number of packets in the packets channel")
+	tlmChannelSizePackets = telemetry.NewGauge("dogstatsd", "packets_channel_size_packets",
```
Can we please give a more descriptive name for this metric? Otherwise it isn't clear how it is different from `packets_channel_size`. Something like:

```suggestion
	tlmChannelSizePackets = telemetry.NewGauge("dogstatsd", "packets_channel_size_by_listener",
```

vickenty commented on comp/dogstatsd/packets/buffer.go:
```diff
@@ -67,15 +71,19 @@ func (pb *Buffer) Append(packet *Packet) {
 		t1 := time.Now()
+
+		TelemetryTrackPackets(pb.packets, pb.listenerID)
+		tlmBufferSizeBytes.Add(-float64(pb.packets.SizeInBytes()), pb.listenerID)
```
The metrics are gauges that are sampled once every 15s, but change many times per second. This is ends up very as a very low sample rate, and doesn't capture at all what happens in between. Worse still if the load also happens at 15s cadence - chances are we'll never even see a random sample of it. Would it make sense to track this as a histogram, or at least min/max/avg set of values?

iksaif replied:
That is a very good point. But I think simple gauges are enough since we mostly care about sustained load scenarios (when the channel is full), and short spikes don't matter much to us (the client should just have higher timeouts)

vickenty commented on comp/dogstatsd/packets/types.go:
```diff
@@ -36,3 +38,20 @@ type Packets []*Packet
+const SizeOfPacket = unsafe.Sizeof(Packet{})
+
+// SizeInBytes returns the size of the packet in bytes
+func (p *Packet) SizeInBytes() int {
```
Would it make sense to call it DataSizeInBytes, to follow the same pattern as the rest of the PR?

iksaif replied:
good idea

vickenty commented on comp/dogstatsd/packets/types.go:
```diff
@@ -36,3 +38,20 @@ type Packets []*Packet
+}
+
+// SizeInBytes returns the size of the packets in bytes
+func (ps *Packets) SizeInBytes() int {
```
Same here

iksaif replied:
good idea

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif commented on pkg/aggregator/context_resolver.go:
```diff
@@ -37,16 +44,40 @@ func (c *Context) release() {
+}
+
 // contextResolver allows tracking and expiring contexts
+// TODO: we should try to separate contextResolvers per origin/listener to get better accounting.
```
Yes we do, but I think we will need a better way to isolate tenants. For sure it comes with challenges, for example metrics with `host:` and no extra k8s cardinality or no origin would still need to go through the same context resolver

vickenty replied:
Let's not leave TODOs that require a lot of context to understand. This should be tracked/captured elsewhere (Jira, RFCs, etc).

iksaif commented on pkg/aggregator/context_resolver.go:
```diff
@@ -37,16 +44,40 @@ func (c *Context) release() {
 }
 
+func (cr *contextResolver) SizeInBytes() int {
+	return ContextSizeInBytes*len(cr.contextsByKey) + int(unsafe.Sizeof(*cr))
```
Ah! I forgot to add telemetry to use this.

Yes the size is imperfect, but this still gives us valuable information (I'll add more warnings in the comments).
Regarding measuring the overhead of the map, it significant but will not change the order of magnitude that we see

iksaif replied:
Yes with a map to a single pointer that is true, `Context` is much bigger than that. I agree that we'll be missing things here, but I'm not sure how we can correct the exposed size.

The thinking is that as-is it still provide good-enough approximation: we can see the estimated usage per pipeline and check, and this already give us a good estimation of what might be using memory which is easier to track over time than profiles

iksaif commented on pkg/util/size.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+// HasSizeInBytes is an interface that can be implemented by any object that has a size in bytes
+type HasSizeInBytes interface {
```
I'm using it to ensure consistency in how we instrument relevant structures. I'd like to eventually use it in agent commands

vickenty replied:
We don't actually use any of the types you apply it to in the same context. Why do they need to conform to the same interface?

How will it be used in the agent commands?

iksaif replied:
The use now (or actually before I start removing one of the existing `DataSizeInBytes()`) was to make sure we would enforce the same way of accounting for these data structures.

The potential use later is to add more introspection command where we'd eventually have a single command that has a `HasSizeInBytes` and can output things with it

vickenty replied:
> or actually before I start removing one of the existing DataSizeInBytes()

This should probably be done in this PR.

iksaif replied:
Ok removing the interface for now

iksaif commented on pkg/util/size.go:
```diff
@@ -0,0 +1,54 @@
+// HasSizeInBytes is an interface that can be implemented by any object that has a size in bytes
+type HasSizeInBytes interface {
+	// SiteInByte Return the size of the object in bytes (not including the size of its content)
+	SiteInByte() int
```
fixed

iksaif commented on pkg/aggregator/internal/tags/store.go:
```diff
@@ -28,7 +30,30 @@ type Entry struct {
+
+// SizeInBytes returns the size of the Entry in bytes.
+func (e Entry) SizeInBytes() int {
+	return e.tags.SizeInBytes() + 8
```
right, removed (initially I though it could be interesting in parts where we list contexts / set of tags, for example the `top` commands)

vickenty replied:
It looks like the new code only uses tagSet#SizeInBytes in one place, here. Can we avoid creating new type and just call util methods from Entry#SizeInBytes?

iksaif replied:
done

iksaif commented on pkg/util/size.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+// SizeOf returns the size of 'v' in bytes (not including its content).
+func SizeOf(v interface{}) int {
```
Removed

iksaif commented on pkg/aggregator/context_resolver.go:
```diff
@@ -37,16 +44,40 @@ func (c *Context) release() {
+		size += cx.DataSizeInBytes()
+	}
+	return size
+}
```
Keeping for now (I'm not using this one for the telemetry because it would not be efficient enough, but I'd like to have it for later )

vickenty replied:
No, please don't add unused code. If you need it later, we can always add it then.

iksaif replied:
Ok, removing for now

iksaif commented on pkg/util/size.go:
```diff
@@ -0,0 +1,54 @@
+// SizeOfString returns the size of a string structure in bytes.
+func SizeOfString() int {
+	return int(StringSize)
+}
```
Right, removing

iksaif committed:
Address comments

iksaif committed:
Remove DataSizeInBytes for contextResolver

iksaif committed:
Add missing slice overhead

vickenty commented on pkg/aggregator/internal/tags/store.go:
```diff
@@ -206,5 +241,7 @@ func (s *entryStats) visit(e *Entry, r uint64) {
 		s.maxSize = n
 	}
 	s.sumSize += n
+	s.sumSizeBytes += e.SizeInBytes() + e.tags.SizeInBytes()
```
Doesn't `e.SizeInBytes` already include `e.tags.SizeInBytes`?

iksaif replied:
It does :(

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif commented on pkg/aggregator/context_resolver.go:
```diff
@@ -82,8 +117,11 @@ func (cr *contextResolver) trackContext(metricSampleContext metrics.MetricSample
 		cr.countsByMtype[mtype]++
+		cr.bytesByMtype[mtype] += uint64(context.SizeInBytes())
 	}
+	tlmContextResolverBytes.Set(float64(cr.SizeInBytes()), cr.id)
```
This is much more efficient than what we're used too (see GetMetricWithLabelValues implementation), but you are right, it is significant.

Without this line:
```
BenchmarkContextResolver1-10          	10419138	       114.9 ns/op
```

With this line:
```
BenchmarkContextResolver1-10          	 6526820	       183.9 ns/op
```

I moved it with the other telemetry metrics for now (even though we could just keep a ref to the gauge instead in the context-resolver)

iksaif replied:
I reworked this part and this should be better. We update less often now and this also removed a bunch of duplicated code.

iksaif commented on pkg/aggregator/context_resolver.go:
```diff
@@ -103,8 +141,10 @@ func (cr *contextResolver) remove(expiredContextKey ckey.ContextKey) {
 
 	if context != nil {
 		cr.countsByMtype[context.mtype]--
+		cr.bytesByMtype[context.mtype] -= uint64(context.SizeInBytes())
```
correct, we want `DataSizeInBytes` here (and the overhead will be the size of the context-resolver itself)

iksaif committed:
aggregator: refactor new metrics

Remove a bunch of accessors in context_resolver and add a
method to update telemetry. Also add a tag to export the kind
of bytes that are counted.

iksaif committed:
Remove context_resolver_bytes for now, it's unreliable

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

vickenty commented on comp/dogstatsd/packets/telemetry.go:
```diff
@@ -52,3 +58,17 @@ func InitTelemetry(buckets []float64) {
+
+// TelemetryUntrackPackets untracks the number of packets in the channel and the number of bytes
+func TelemetryUntrackPackets(packets Packets) {
+	for _, packet := range packets {
```
Would it make sense for this function to use Packets#SizeInBytes, like the `TelemetryTrackPackets`?

iksaif replied:
This could not work because packets are not guaranteed to belong to the same listenerId.

vickenty replied:
Why not? Isn't it exactly the same `packets` object we handle in `TelemetryTrackPackets`, just on the other side of the channel?

iksaif replied:
`Packets` is just a slice of `Packet`
- when we call `TelemetryTrackPackets(packets, listenerId)` the caller tells us that it's known that all sub-packet have the same value for `listenerId`
- when we call `TelemetryUnTrackPackets(packets)` something might have merged `Packets` together and we don't know about it

vickenty commented on pkg/aggregator/context_resolver.go:
```diff
@@ -103,21 +134,47 @@ func (cr *contextResolver) remove(expiredContextKey ckey.ContextKey) {
+		// Limit un-needed cardinality (especially because each check has its own resolver)
+		gauge := countsByMTypeGauge.WithValues(cr.id, mtype)
+		// If the gauge doesn't exist, and we have nothing to report, continue
+		if gauge == nil && count == 0 {
```
`WithValues` [never returns nil](https://github.com/DataDog/datadog-agent/blob/main/comp/core/telemetry/prom_gauge.go#L49).

iksaif replied:
Ah right, I was thinking `GetMetricWithLabelValues` .. fixing

vickenty commented on pkg/aggregator/internal/tags/store.go:
```diff
@@ -28,9 +30,36 @@ type Entry struct {
+
+// DataSizeInBytes returns the size of the tagSet data in bytes.
+func (t tagSet) DataSizeInBytes() int {
+	// This could be optimized by keeping track of the size when the TagSet is created.
```
As an aside, if you want to optimize this, you could track of bytes in the store as entries are added and removed, instead of recomputing in `visit`. This way there is no need to store extra data in the `Entry` or the `tagSet`.

iksaif replied:
I looked at the code again, and this seems cheap enough currently to leave it this way no?
I was trying to keep the telemetry code in a single place, as it is today

iksaif committed:
Merge remote-tracking branch 'origin/main' into corentin.chary/instrument-memory

iksaif committed:
Simplify code and fix `gauge` check

iksaif commented:
/merge
